File: app/src/main/java/com/checkin/app/checkin/Manager/ManagerRepository.java
package com.checkin.app.checkin.Manager;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.checkin.app.checkin.Data.ApiClient;
import com.checkin.app.checkin.Data.ApiResponse;
import com.checkin.app.checkin.Data.BaseRepository;
import com.checkin.app.checkin.Data.NetworkBoundResource;
import com.checkin.app.checkin.Data.Resource;
import com.checkin.app.checkin.Data.RetrofitLiveData;
import com.checkin.app.checkin.Data.WebApiService;
import com.checkin.app.checkin.Manager.Model.ManagerSessionInvoiceModel;
import com.checkin.app.checkin.Manager.Model.ManagerStatsModel;
import com.checkin.app.checkin.Misc.GenericDetailModel;
import com.checkin.app.checkin.session.model.CheckoutStatusModel;
import com.checkin.app.checkin.session.model.QRResultModel;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class ManagerRepository extends BaseRepository {
    private static ManagerRepository INSTANCE;
    private final WebApiService mWebService;

    private ManagerRepository(Context context) {
        mWebService = ApiClient.getApiService(context);
    }

    public static ManagerRepository getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (ManagerRepository.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ManagerRepository(application.getApplicationContext());
                }
            }
        }
        return INSTANCE;
    }

    public LiveData<Resource<ManagerStatsModel>> getManagerStats(long restaurantId) {
        return new NetworkBoundResource<ManagerStatsModel, ManagerStatsModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ManagerStatsModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.getRestaurantManagerStats(restaurantId));
            }

            @Override
            protected void saveCallResult(ManagerStatsModel data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<CheckoutStatusModel>> manageSessionCheckout(long sessionId) {
        return new NetworkBoundResource<CheckoutStatusModel, CheckoutStatusModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<CheckoutStatusModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.putSessionCheckout(sessionId));
            }

            @Override
            protected void saveCallResult(CheckoutStatusModel data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ManagerSessionInvoiceModel>> getManagerSessionInvoice(long sessionId) {
        return new NetworkBoundResource<ManagerSessionInvoiceModel, ManagerSessionInvoiceModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ManagerSessionInvoiceModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.getManagerSessionInvoice(sessionId));
            }

            @Override
            protected void saveCallResult(ManagerSessionInvoiceModel data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<GenericDetailModel>> putManageSessionBill(long sessionId, ObjectNode data) {
        return new NetworkBoundResource<GenericDetailModel, GenericDetailModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<GenericDetailModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.putManageSessionBill(sessionId, data));
            }

            @Override
            protected void saveCallResult(GenericDetailModel data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<QRResultModel>> managerInitiateSession(ObjectNode requestJson) {
        return new NetworkBoundResource<QRResultModel, QRResultModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<QRResultModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.postManageInitiateSession(requestJson));
            }

            @Override
            protected void saveCallResult(QRResultModel data) {
                //saveCallResult code
            }
        }.getAsLiveData();
    }
}

File: app/src/main/java/com/checkin/app/checkin/manager/ManagerRepository.kt
package com.checkin.app.checkin.manager

import android.app.Application
import android.content.Context
import androidx.lifecycle.LiveData
import com.checkin.app.checkin.data.BaseRepository
import com.checkin.app.checkin.data.Converters
import com.checkin.app.checkin.data.network.ApiClient.Companion.getApiService
import com.checkin.app.checkin.data.network.ApiResponse
import com.checkin.app.checkin.data.network.RetrofitLiveData
import com.checkin.app.checkin.data.network.WebApiService
import com.checkin.app.checkin.data.resource.NetworkBoundResource
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.manager.models.*
import com.checkin.app.checkin.misc.models.GenericDetailModel
import com.checkin.app.checkin.session.models.CheckoutStatusModel
import com.checkin.app.checkin.session.models.QRResultModel
import com.checkin.app.checkin.utility.SingletonHolder
import com.fasterxml.jackson.databind.node.ObjectNode

class ManagerRepository private constructor(context: Context) : BaseRepository() {
    private val mWebService: WebApiService = getApiService(context)

    fun getManagerStats(restaurantId: Long): LiveData<Resource<ManagerStatsModel>> {
        return object : NetworkBoundResource<ManagerStatsModel, ManagerStatsModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ManagerStatsModel>> {
                return RetrofitLiveData(mWebService.getRestaurantManagerStats(restaurantId))
            }
        }.asLiveData
    }

    fun manageSessionCheckout(sessionId: Long): LiveData<Resource<CheckoutStatusModel>> {
        return object : NetworkBoundResource<CheckoutStatusModel, CheckoutStatusModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<CheckoutStatusModel>> {
                return RetrofitLiveData(mWebService.putSessionCheckout(sessionId))
            }
        }.asLiveData
    }

    fun getManagerSessionInvoice(sessionId: Long): LiveData<Resource<ManagerSessionInvoiceModel>> {
        return object : NetworkBoundResource<ManagerSessionInvoiceModel, ManagerSessionInvoiceModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ManagerSessionInvoiceModel>> {
                return RetrofitLiveData(mWebService.getManagerSessionInvoice(sessionId))
            }
        }.asLiveData
    }

    fun putManageSessionBill(sessionId: Long, data: ObjectNode): LiveData<Resource<GenericDetailModel>> {
        return object : NetworkBoundResource<GenericDetailModel, GenericDetailModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<GenericDetailModel>> {
                return RetrofitLiveData(mWebService.putManageSessionBill(sessionId, data))
            }
        }.asLiveData
    }

    fun managerInitiateSession(requestJson: ObjectNode): LiveData<Resource<QRResultModel>> {
        return object : NetworkBoundResource<QRResultModel, QRResultModel>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<QRResultModel>> {
                return RetrofitLiveData(mWebService.postManageInitiateSession(requestJson))
            }
        }.asLiveData
    }

    fun managerSessionSwitchTable(sessionId: Long, requestJson: ObjectNode): LiveData<Resource<ObjectNode>> {
        return object : NetworkBoundResource<ObjectNode, ObjectNode>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<ObjectNode>> {
                return RetrofitLiveData(mWebService.postTableSwitch(sessionId, requestJson))
            }
        }.asLiveData
    }

    fun getRestaurantScheduledSession(restaurantId: Long): LiveData<Resource<List<ShopScheduledSessionModel>>> {
        return object : NetworkBoundResource<List<ShopScheduledSessionModel>, List<ShopScheduledSessionModel>>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<List<ShopScheduledSessionModel>>> {
                return RetrofitLiveData(mWebService.getScheduledSessionsById(restaurantId))
            }
        }.asLiveData
    }

    fun getScheduledSessionDetail(sessionId: Long): LiveData<Resource<ShopScheduledSessionDetailModel>> {
        return object : NetworkBoundResource<ShopScheduledSessionDetailModel, ShopScheduledSessionDetailModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<ShopScheduledSessionDetailModel>> {
                return RetrofitLiveData(mWebService.getManageScheduledSessionDetail(sessionId))
            }

        }.asLiveData
    }

    fun acceptScheduledSession(sessionId: Long, data: PreparationTimeModel): LiveData<Resource<PreparationTimeModel>> {
        return object : NetworkBoundResource<PreparationTimeModel, PreparationTimeModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<PreparationTimeModel>> {
                return RetrofitLiveData(mWebService.patchManageScheduledSessionAccept(sessionId, data))
            }
        }.asLiveData
    }

    fun doneScheduledSession(sessionId: Long): LiveData<Resource<ScheduledSessionDoneModel>> {
        return object : NetworkBoundResource<ScheduledSessionDoneModel, ScheduledSessionDoneModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<ScheduledSessionDoneModel>> {
                return RetrofitLiveData(mWebService.patchManageScheduledSessionDone(sessionId, Converters.objectMapper.createObjectNode()))
            }
        }.asLiveData
    }

    fun rejectScheduledSession(sessionId: Long, data: ScheduledSessionCancelModel): LiveData<Resource<GenericDetailModel>> {
        return object : NetworkBoundResource<GenericDetailModel, GenericDetailModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<GenericDetailModel>> {
                return RetrofitLiveData(mWebService.postManageScheduledSessionReject(sessionId, data))
            }
        }.asLiveData
    }

    companion object : SingletonHolder<ManagerRepository, Application>({ ManagerRepository(it.applicationContext) })
}